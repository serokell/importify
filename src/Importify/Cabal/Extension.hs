-- | Functions that work with 'Extension's in .cabal file and in
-- separate.

module Importify.Cabal.Extension
       ( buildInfoExtensions
       , withHarmlessExtensions
       ) where

import           Universum

--import qualified Distribution.ModuleName         as Cabal
import           Distribution.PackageDescription (BuildInfo (..))
import qualified Language.Haskell.Extension      as Cabal (Extension (..))
import           Language.Haskell.Exts.Extension (Extension (..), KnownExtension (..))


-- | Get list of all extensions from 'BuildInfo' and convert them into
-- 'HSE.Extension'.
buildInfoExtensions :: BuildInfo -> [Extension]
buildInfoExtensions BuildInfo{..} = mapMaybe cabalExtToHseExt
                                  $ defaultExtensions ++ otherExtensions

cabalExtToHseExt :: Cabal.Extension -> Maybe Extension
cabalExtToHseExt = readMaybe . show

{-
showExt :: Cabal.Extension -> String
showExt (Cabal.EnableExtension ext)   = show ext
showExt (Cabal.DisableExtension ext)  = "No" ++ show ext
showExt (Cabal.UnknownExtension name) = name
-}

-- | This function add list of harmless extensions wich helps to avoid
-- some parsing errors but doesn't affect already correct parsing
-- results.
withHarmlessExtensions :: [Extension] -> [Extension]
withHarmlessExtensions = ordNub
                       . (++ map EnableExtension [ MultiParamTypeClasses
                                                 , FlexibleContexts
                                                 , ConstraintKinds
                                                 , ExplicitNamespaces
                                                 ])
