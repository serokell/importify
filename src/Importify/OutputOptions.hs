-- | Contains data type which specifies output options for @importify@ commands.

module Importify.OutputOptions
       ( OutputOptions (..)
       , printWithOutputOptions
       ) where

import           Universum

-- | This data type dictates how output of @importify@ should be outputed.
data OutputOptions = ToConsole        -- ^ Print to console
                   | ToFile FilePath  -- ^ Print to specified file
                   deriving (Show)

-- | Prints text using given 'OutputOptions'.
printWithOutputOptions :: OutputOptions -> Text -> IO ()
printWithOutputOptions ToConsole   text = putText text
printWithOutputOptions (ToFile to) text = writeFile to text
