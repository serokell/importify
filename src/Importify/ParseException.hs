-- | This module contains data type for parsing exception and pretty
-- formatting of such exceptions.

module Importify.ParseException
       ( ModuleParseException (..)
       , eitherParseResult
       , reportErrorsIfAny
       , setMpeFile
       ) where

import           Universum

import           Fmt                   (blockListF, build, indentF, (+|), (|+))
import           Language.Haskell.Exts (ParseResult (..), SrcLoc (srcFilename),
                                        prettyPrint)

import           Extended.Data.Str     (charWrap, wordWrap)
import           Extended.System.Wlog  (printWarning)

data ModuleParseException = MPE !SrcLoc !String
    deriving (Show)

instance Exception ModuleParseException
instance Buildable ModuleParseException where
    build (MPE loc reason) = "Location:\n"
                          +| indentF 4 (build $ charWrap 80 $ prettyPrint loc)
                          |+ "Reason:\n"
                          +| indentF 4 (build $ wordWrap 80 reason)
                          |+ ""

-- | Updates file name of error location. Sometimes error location
-- looks like @- Location: : -1: -1@ which is not very helpful.
setMpeFile :: FilePath -> ModuleParseException -> ModuleParseException
setMpeFile modulePath (MPE loc msg) = MPE (loc {srcFilename = modulePath}) msg

-- | Converts 'ParseResult' into 'Either' making error look pretty.
eitherParseResult :: ParseResult res
                  -> Either ModuleParseException res
eitherParseResult (ParseOk res)            = Right res
eitherParseResult (ParseFailed loc reason) = Left $ MPE loc reason

-- | Pretty printing 'NonEmpty' list of errors in really nice way.
prettyParseErrors :: Text -> NonEmpty ModuleParseException -> Text
prettyParseErrors libName exceptions =
    "Next errors occured during caching of package: "+|libName|+"\n"
 +| indentF 4 (blockListF exceptions) |+ ""

-- | Prints parse errors if list of errors is not empty.
reportErrorsIfAny :: [ModuleParseException] -> Text -> IO ()
reportErrorsIfAny exceptions libName = whenNotNull exceptions $
    printWarning . prettyParseErrors libName
