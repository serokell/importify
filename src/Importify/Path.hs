{-# LANGUAGE ExplicitForAll #-}
{-# LANGUAGE QuasiQuotes    #-}
{-# LANGUAGE TypeOperators  #-}

-- | This module contains common utilities for working with importify cache.

module Importify.Path
       ( -- * Predefined directories
         importifyDir
       , importifyPath
       , extensionsFile
       , extensionsPath
       , modulesFile
       , modulesPath
       , symbolsDir
       , symbolsPath

         -- * Paths for tests
       , testDataPath
       , testUnusedPath

         -- * Utility functions to work with files and directories
       , decodeFileOrMempty
       , doInsideDir
       , findCabalFile
       , lookupToRoot
       ) where

import           Universum

import           Data.Aeson           (FromJSON, eitherDecodeStrict)
import qualified Data.ByteString      as BS (readFile)
import           Fmt                  ((+|), (+||), (|+), (||+))
import           Path                 (Abs, Dir, File, Path, Rel, dirname, fromAbsDir,
                                       fromAbsFile, fromRelDir, fromRelFile, parent,
                                       reldir, relfile, toFilePath, (</>))
import           Path.IO              (doesFileExist, ensureDir, getCurrentDir, listDir)
import           System.FilePath      (takeExtension)
import           Turtle               (cd, pwd)

import           Extended.System.Wlog (printNotice, printWarning)

importifyPath :: Path Rel Dir
importifyPath = [reldir|.importify/|]

-- | Path to file that stores mapping from module names to their packages.
modulesPath :: Path Rel File
modulesPath = [relfile|modules|]

symbolsPath :: Path Rel Dir
symbolsPath = [reldir|symbols/|]

-- | Path to golden tests.
testDataPath :: Path Rel Dir
testDataPath = [reldir|test/test-data/|]

-- | Path to tests for removing unused imports.
testUnusedPath :: Path Rel Dir
testUnusedPath = testDataPath </> [reldir|unused|]

-- | Path to JSON-encoded Map from target to its list of default extensions.
extensionsPath :: Path Rel File
extensionsPath = [relfile|extensions|]

importifyDir, extensionsFile, modulesFile, symbolsDir :: FilePath
importifyDir   = fromRelDir  importifyPath
extensionsFile = fromRelFile extensionsPath
modulesFile    = fromRelFile modulesPath
symbolsDir     = fromRelDir  symbolsPath

-- | Returns relative path to cabal file under given directory.
findCabalFile :: MonadIO m => Path Abs Dir -> m $ Maybe $ Path Abs File
findCabalFile projectPath = do
    (_, projectFiles) <- listDir projectPath
    return $ find isCabalFile projectFiles

isCabalFile :: Path Abs File -> Bool
isCabalFile = (== ".cabal") . takeExtension . fromAbsFile

-- | Create given directory and perform given action inside it.
doInsideDir :: (MonadIO m, MonadMask m) => Path Abs Dir -> m a -> m a
doInsideDir dir action = do
    thisDirectory <- pwd
    bracket_ (do ensureDir dir
                 cd $ fromString $ fromAbsDir dir)
             (cd thisDirectory)
             action

-- | Walk up till root while unpure predicate is 'False'. Returns
-- absolute path to directory where predicate is 'True' and suffix of
-- current directory prepended to given file.
lookupToRoot :: (Path Abs Dir -> IO Bool)
             -> Path Rel File
             -> IO (Maybe (Path Abs Dir, Path Rel File))
lookupToRoot predicate relativeFile = do
    currentDir <- getCurrentDir
    pathLoop currentDir relativeFile
  where
    pathLoop :: Path Abs Dir -> Path Rel File -> IO (Maybe (Path Abs Dir, Path Rel File))
    pathLoop directory file = do
        predicateIsTrue <- predicate directory
        if predicateIsTrue then
            return $ Just (directory, file)
        else if parent directory == directory then  -- fixpoint reached
            return Nothing
        else do
            let parentDir = parent  directory
            let   thisDir = dirname directory
            pathLoop parentDir (thisDir </> file)

-- | Tries to read file and then 'decode' it. If either of two phases
-- fails then 'mempty' returned and warning is printed to console.
decodeFileOrMempty :: forall t m f b .
                      (FromJSON t, Monoid m, MonadIO f)
                   => Path b File  -- ^ Path to json data
                   -> (t -> f m)   -- ^ Action from decoded value
                   -> f m
decodeFileOrMempty file onDecodedContent = do
    isFileExist <- doesFileExist file

    if isFileExist then
        eitherDecodeStrict <$> liftIO (BS.readFile $ toFilePath file) >>= \case
            Right value -> onDecodedContent value
            Left msg    -> do
              let warning = "File '"+||file||+"' decoded incorrectly because of: "+|msg|+""
              mempty <$ printWarning warning
    else do
        let msg = "File '"+||file||+
                  "' doesn't exist: caching first time or previous caching failed"
        mempty <$ printNotice msg
