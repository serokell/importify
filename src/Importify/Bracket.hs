-- | This module contains functions which wrap @importify@ commands.

module Importify.Bracket
       ( ImportifyArguments     (..)
       , ImportifyFileException (..)
       , ImportifyFunction
       , ImportifyResult
       , importifyAstBracket
       , importifyOptionsBracket
       , importifyPathBracket
       , loadEnvironment
       ) where

import           Universum

import qualified Data.HashMap.Strict          as HM
import qualified Data.Map                     as M

import           Fmt                          (fmt, (+|), (|+))
import           Language.Haskell.Exts        (Comment (..), Extension, ImportDecl,
                                               Module (..), ModuleName (..), SrcSpanInfo,
                                               exactPrint, parseExtension,
                                               parseFileContentsWithComments)
import           Language.Haskell.Exts.Parser (ParseMode (..), defaultParseMode)
import           Language.Haskell.Names       (Environment, loadBase, readSymbols)
import           Path                         (Abs, File, Path, fromAbsFile, fromRelFile,
                                               parseRelDir, parseRelFile, (</>))
import           Path.IO                      (doesDirExist, getCurrentDir)

import           Extended.System.Wlog         (printError, printNotice)
import           Importify.Cabal              (ExtensionsMap, ModulesBundle (..),
                                               ModulesMap, TargetId, targetIdDir)
import           Importify.OutputOptions
import           Importify.ParseException     (eitherParseResult, setMpeFile)
import           Importify.Path               (decodeFileOrMempty, doInsideDir,
                                               extensionsPath, importifyPath,
                                               lookupToRoot, modulesPath, symbolsPath)
import           Importify.Pretty
import           Importify.Syntax

-- | Type that represents exception occurred during @importify@ command.
newtype ImportifyFileException = IFE Text

-- | Convenient type alias for return result of @importify file@ command.
type ImportifyResult = Either ImportifyFileException Text

-- | Bracket for every command which process file.
importifyOptionsBracket :: (Path Abs File -> IO ImportifyResult)
                        -> OutputOptions
                        -> FilePath
                        -> IO ()
importifyOptionsBracket importify options srcFile = do
    srcPath   <- parseRelFile srcFile
    foundRoot <- lookupToRoot (doesDirExist . (</> importifyPath)) srcPath
    case foundRoot of
        Nothing ->
            printError "Directory '.importify' is not found. Either cache for project \
                       \is not created or not running from project directory."
        Just (rootDir, srcFromRootPath) -> do
            curDir          <- getCurrentDir
            importifyResult <- doInsideDir rootDir $ importify $ curDir </> srcFromRootPath
            handleOptions importifyResult
  where
    handleOptions :: Either ImportifyFileException Text -> IO ()
    handleOptions (Left (IFE msg))    = printError msg
    handleOptions (Right modifiedSrc) = printWithOutputOptions options modifiedSrc

-- | All needed data for @importify@ command which processes one file.
data ImportifyArguments = ImportifyArguments
    { importifyArgumentsAst        :: !(Module SrcSpanInfo)
    , importifyArgumentsModulesMap :: !ModulesMap
    , importifyArgumentsSrc        :: !Text
    , importifyArgumentsComments   :: ![Comment]
    } deriving (Show)

-- | Runs given action over parsed AST.
importifyPathBracket :: (ImportifyArguments -> IO ImportifyResult)
                     -> Path Abs File
                     -> IO ImportifyResult
importifyPathBracket importify srcPath = do
    let srcFile = fromAbsFile srcPath

    modulesMap <- readModulesMap
    extensions <- readExtensions srcPath modulesMap

    whenNothing_ (HM.lookup (fromAbsFile srcPath) modulesMap) $
        printNotice $ "File '"+|srcFile|+"' is not cached: new file or caching error"

    src <- readFile srcFile
    let parseResult = eitherParseResult
                    $ parseFileContentsWithComments (defaultParseMode { extensions = extensions })
                    $ toString src

    case parseResult of
        Left exception       -> return $ Left $ IFE $ setMpeFile srcFile exception |+ ""
        Right (ast,comments) -> importify ImportifyArguments
                                    { importifyArgumentsAst        = ast
                                    , importifyArgumentsModulesMap = modulesMap
                                    , importifyArgumentsSrc        = src
                                    , importifyArgumentsComments   = comments
                                    }

-- | Type of functions which processes imports.
type ImportifyFunction = Module SrcSpanInfo           -- ^ Module where symbols should be removed
                      -> Environment                  -- ^ Environment of all cached modules
                      -> [ImportDecl SrcSpanInfo]     -- ^ Imports from module
                      -> IO [ImportDecl SrcSpanInfo]  -- ^ New imports

-- | Process imports in AST with given function.
-- TODO: this function is ugly :( Write it nicer somehow...
importifyAstBracket :: ImportifyFunction
                    -> ImportifyArguments
                    -> IO ImportifyResult
importifyAstBracket importify ImportifyArguments{..}
    | ast@(Module _ _ _ imports _) <- importifyArgumentsAst =
        Right <$> case importSlice imports of
            Nothing           -> return importifyArgumentsSrc
            Just (start, end) -> importifyAst start end ast imports
    | otherwise = return $ Left $ IFE "Module wasn't parsed correctly"
  where
    importifyAst :: Int
                 -> Int
                 -> Module SrcSpanInfo
                 -> [ImportDecl SrcSpanInfo]
                 -> IO Text
    importifyAst start end ast imports = do
        let codeLines        = lines importifyArgumentsSrc
        let (preamble, rest) = splitAt (start - 1) codeLines
        let (impText, decls) = splitAt (end - start + 1) rest

        environment       <- loadEnvironment importifyArgumentsModulesMap
        newImports        <- importify ast environment imports
        let printedImports = printLovelyImports start
                                                end
                                                importifyArgumentsComments
                                                impText
                                                newImports

        return $ unlines preamble
              <> unlines printedImports
              <> unlines decls

-- | Reads 'ModulesMap' from @.importify/modules@.
readModulesMap :: IO ModulesMap
readModulesMap = decodeFileOrMempty (importifyPath </> modulesPath) pure

readExtensions :: Path Abs File -> ModulesMap -> IO [Extension]
readExtensions srcPath modulesMap = do
    case HM.lookup (fromAbsFile srcPath) modulesMap of
        Nothing                -> return []
        Just ModulesBundle{..} -> do
            packagePath <- parseRelDir $ toString mbPackage
            projectPath <- getCurrentDir
            let pathToExtensions = projectPath
                               </> importifyPath
                               </> symbolsPath
                               </> packagePath
                               </> extensionsPath

            let lookupExtensions = fromMaybe [] . getExtensions mbTarget
            decodeFileOrMempty @ExtensionsMap
                               pathToExtensions
                               (return . lookupExtensions)

getExtensions :: TargetId -> ExtensionsMap -> Maybe [Extension]
getExtensions targetId = fmap (map parseExtension) . HM.lookup targetId

loadEnvironment :: ModulesMap -> IO Environment
loadEnvironment modulesMap = do
    baseEnvironment <- loadBase

    let moduleBundles = HM.elems modulesMap
    packages <- forM moduleBundles $ \ModulesBundle{..} -> do
        packagePath     <- parseRelDir  $ toString mbPackage
        symbolsFilePath <- parseRelFile $ mbModule ++ ".symbols"
        targetPath      <- parseRelDir $ toString $ targetIdDir mbTarget
        let pathToSymbols = importifyPath
                        </> symbolsPath
                        </> packagePath
                        </> targetPath
                        </> symbolsFilePath
        moduleSymbols <- readSymbols (fromRelFile pathToSymbols)
        pure (ModuleName () mbModule, moduleSymbols)

    return $ M.union baseEnvironment (M.fromList packages)
