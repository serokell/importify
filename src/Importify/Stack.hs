{-# LANGUAGE ExplicitForAll      #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell     #-}

-- | Utilities which allows to use @stack@ tools for different
-- dependencies stuff.

module Importify.Stack
       ( QueryPackage   (..)
       , LocalPackages  (..)
       , RemotePackages (..)

       , ghcIncludePath
       , pkgName
       , stackListDependencies
       , stackListPackages
       , stackProjectRoot
       , upgradeWithVersions
       ) where

import           Universum

import qualified Control.Foldl        as Fold (head, list)
import qualified Data.HashMap.Strict  as HM
import           Data.List            (partition)
import qualified Data.Text            as T
import           Data.Yaml            (FromJSON (parseJSON), Value (Object),
                                       decodeEither', prettyPrintParseException,
                                       withObject, (.:))
import           Path                 (Abs, Dir, Path, PathException, dirname, fromAbsDir,
                                       mkRelDir, parent, parseAbsDir, (</>))
import           Path.IO              (doesDirExist)
import           System.FilePath      (splitPath)
import           Turtle               (Line, Shell, inproc, lineToText, linesToText)
import qualified Turtle               (fold)

import           Extended.System.Wlog (printWarning)

shStack :: [Text] -> Shell Line
shStack args = inproc "stack" args empty

pathArgs, rootArgs, depsArgs :: [Text]
pathArgs = ["path", "--compiler-bin"]
rootArgs = ["path", "--project-root"]
depsArgs = ["list-dependencies", "--test", "--bench"]

-- | This function finds path to directory where @include@ for ghc lies.
-- Filepath looks like this:
-- @
--   ~/.stack/programs/x86_64-linux/ghc-8.0.2/lib/ghc-8.0.2/include
-- @
-- This function needed to tell dependencies about files like @"MachDeps.h"@.
--
-- TODO: use GHC path from project?
ghcIncludePath :: MaybeT IO (Path Abs Dir)
ghcIncludePath = do
    ghcBinLine <- MaybeT $ Turtle.fold (shStack pathArgs) Fold.head

    -- ghcBinText ≡ /home/user/.stack/programs/x86_64-linux/ghc-8.0.2/bin
    ghcBinText    <- parseAbsDir $ toString $ lineToText ghcBinLine
    let ghcProgram = parent ghcBinText   -- w/o bin
    let ghcName    = dirname ghcProgram  -- ≡ ghc-8.0.2

    -- ghcInclude ≡ /home/user/.stack/programs/x86_64-linux/ghc-8.0.2/lib/ghc-8.0.2/include
    let ghcIncludeDir = ghcProgram
                    </> $(mkRelDir "lib/")
                    </> ghcName
                    </> $(mkRelDir "include/")

    guardM $ doesDirExist ghcIncludeDir
    return ghcIncludeDir

-- | Acquires project using @stack path --project-root@ command.
stackProjectRoot :: MaybeT IO (Path Abs Dir)
stackProjectRoot = do
    projectRootLine <- MaybeT $ Turtle.fold (shStack rootArgs) Fold.head
    let projectRoot = lineToText projectRootLine
    if ".stack/global-project" `T.isSuffixOf` projectRoot then do
        printWarning "importify was executed outside of project" *> empty
    else case eitherParseRoot projectRoot of
        Left exception        -> printWarning (show exception) *> empty
        Right projectRootPath -> return projectRootPath
  where
    eitherParseRoot :: Text -> Either SomeException (Path Abs Dir)
    eitherParseRoot = parseAbsDir . toString

-- | Extract all dependencies with versions using
-- @stack list-dependencies@ shell command.
stackListDependencies :: MonadIO m => m (HashMap Text Text)
stackListDependencies = do
    dependencies   <- Turtle.fold (shStack depsArgs) Fold.list
    let wordifyDeps = map (words . lineToText) dependencies
    let pairifyDeps = pairifyList wordifyDeps
    return $ HM.fromList pairifyDeps
  where
    pairifyList :: [[a]] -> [(a,a)]
    pairifyList        []  = []
    pairifyList ([x,y]:xs) = (x,y) : pairifyList xs
    pairifyList     (_:xs) =         pairifyList xs

-- | Takes mapping from package names to their versions and list of
-- packages adding version to each package which is inside dictionary.
upgradeWithVersions :: HashMap Text Text -> [Text] -> [Text]
upgradeWithVersions versions = go
  where
    go        []  = []
    go (lib:libs) = case HM.lookup lib versions of
        Nothing      -> lib                   : go libs
        Just version -> lib <> "-" <> version : go libs

-- | Queries list of all local packages for project. If some errors
-- occur then warning is printed into console and empty list returned.
stackListPackages :: forall m . (MonadIO m, MonadCatch m)
                  => m (LocalPackages, RemotePackages)
stackListPackages = do
    pkgsYaml    <- linesToText <$> Turtle.fold (shStack ["query"]) Fold.list
    let parseRes = decodeEither' $ encodeUtf8 pkgsYaml
    case parseRes of
        Left exception -> do
            printWarning $ toText $ prettyPrintParseException exception
            return mempty
        Right (StackQueryResult packages) -> do
            localPackages <- mapM toPackage packages `catch` parseErrorHandler
            let (locals, remotes) = partition (isLocalPackage . qpPath) localPackages
            return (LocalPackages locals, RemotePackages remotes)
  where
    toPackage :: (Text, (FilePath, Text)) -> m QueryPackage
    toPackage (qpName, (path, qpVersion)) = do
        qpPath <- parseAbsDir path
        return QueryPackage{..}

    parseErrorHandler :: PathException -> m [QueryPackage]
    parseErrorHandler exception =
        [] <$ printWarning ("'stack query' exception: " <> show exception)

    isLocalPackage :: Path Abs Dir -> Bool
    isLocalPackage = notElem ".stack-work/" . splitPath . fromAbsDir

-- | This data type represents package returned by @stack query@ command.
data QueryPackage = QueryPackage
    { qpName    :: Text          -- ^ @importify@
    , qpPath    :: Path Abs Dir  -- ^ @\/home\/user\/importify\/@
    , qpVersion :: Text          -- ^ 1.0
    } deriving (Eq, Show)

-- | Show full name of 'QueryPackage' with version.
pkgName :: QueryPackage -> Text
pkgName QueryPackage{..} = qpName <> "-" <> qpVersion

-- | Local subpackages from exactly this project.
newtype LocalPackages = LocalPackages [QueryPackage]
    deriving (Eq, Monoid)

-- | Remote packages, from GitHub or other locations.
newtype RemotePackages = RemotePackages [QueryPackage]
    deriving (Eq, Monoid)

newtype StackQueryResult = StackQueryResult [(Text, (FilePath, Text))]
    deriving Show

instance FromJSON StackQueryResult where
    parseJSON = withObject "stack query" $ \obj -> do
        Just (Object locals) <- pure $ HM.lookup "locals" obj
        packages <- forM locals $ withObject "package" $ \pkgObj -> do
            pkgPath    :: FilePath <- pkgObj .: "path"
            pkgVersion :: Text     <- pkgObj .: "version"
            pure (pkgPath, pkgVersion)
        pure $ StackQueryResult $ HM.toList packages
