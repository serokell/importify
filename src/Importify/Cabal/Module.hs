-- | Functions to operate with @exposed-modules@ and @other-modules@
-- parts of .cabal file.

module Importify.Cabal.Module
       ( modulePaths
       , splitOnExposedAndOther
       ) where

import           Universum                       hiding (fromString)

import           Data.List                       (partition)
import           Distribution.ModuleName         (ModuleName, toFilePath)
import qualified Distribution.ModuleName         as Cabal
import           Distribution.PackageDescription (BuildInfo (..), Library (..))
import qualified Language.Haskell.Exts           as HSE
import           Path                            (Abs, Dir, File, Path, Rel, parseRelDir,
                                                  parseRelFile, (</>))
import           Path.IO                         (doesFileExist)

import           Importify.Syntax                (getModuleTitle)

-- | Split list of modules into /exposed/ modules and /other/ modules for given library.
-- __First__ element of pair represents /exposed/ modules.
-- __Second__ element of paris represents /other/ modules.
splitOnExposedAndOther :: Library
                       -> [HSE.Module l]
                       -> ([HSE.Module l], [HSE.Module l])
splitOnExposedAndOther Library{..} =
    partition ((`elem` exposedModules) . Cabal.fromString . getModuleTitle)

-- | Returns list of absolute paths to all modules inside given target.
modulePaths :: Path Abs Dir
            -- ^ Absolute path to project directory
            -> BuildInfo
            -- ^ 'BuildInfo' for given target
            -> Either [ModuleName] FilePath
            -- ^ Modules for Library and path for others
            -> IO [Path Abs File]
modulePaths packagePath BuildInfo{..} extra = do
    let (cur, others) = partition (== ".") hsSourceDirs
    case (cur, others) of
        (_here,    []) -> collectModulesHere
        ([]   , paths) -> collectModulesThere paths
        (_here, paths) -> liftA2 (++) collectModulesHere (collectModulesThere paths)
  where
    modulesToPaths :: [ModuleName] -> IO [Path Rel File]
    modulesToPaths = mapM (parseRelFile . (++ ".hs") . toFilePath)

    targetModulePaths :: IO [Path Rel File]
    targetModulePaths = case extra of
        Left modules -> modulesToPaths $ otherModules ++ modules
        Right path   -> liftA2 (:) (parseRelFile path) (modulesToPaths otherModules)

    addDir :: Path Abs Dir -> [Path Rel File] -> [Path Abs File]
    addDir dir = map (dir </>)

    collectModulesHere :: IO [Path Abs File]
    collectModulesHere = do
        paths <- targetModulePaths
        let packagePaths = addDir packagePath paths
        keepExistingModules packagePaths

    collectModulesThere :: [FilePath] -> IO [Path Abs File]
    collectModulesThere dirs = do
        dirPaths <- mapM parseRelDir dirs
        modPaths <- targetModulePaths
        concatForM dirPaths $ \dir ->
          keepExistingModules $ addDir (packagePath </> dir) modPaths

    keepExistingModules :: [Path Abs File] -> IO [Path Abs File]
    keepExistingModules = filterM doesFileExist
