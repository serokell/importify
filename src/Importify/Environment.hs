{-# LANGUAGE ConstraintKinds        #-}
{-# LANGUAGE FlexibleContexts       #-}
{-# LANGUAGE FlexibleInstances      #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE MultiParamTypeClasses  #-}
{-# LANGUAGE TemplateHaskell        #-}

-- | This module contains enrinment for @importify cache@ command
-- which is running inside @ReaderT env IO@ monad.

module Importify.Environment
       ( -- * Base monad for @importify cache@ command
         RIO (..)

         -- * Environment for @importify cache@ command
       , CacheEnvironment (..)
       , HasGhcIncludeDir
       , HasPathToImportify
       , HasSaveSources
       , ghcIncludeDir
       , pathToImportify
       , pathToSymbols
       , saveSources

         -- * Runner for cache commend
       , runCache
       ) where

import           Universum

import           Lens.Micro.Platform (SimpleGetter, makeLensesWith, to)
import           Path                (Abs, Dir, Path, (</>))
import           Path.IO             (getCurrentDir)

import           Extended.Lens.TH    (fieldsVerboseLensRules)
import           Importify.Path      (doInsideDir, importifyPath, symbolsPath)
import           Importify.Stack     (ghcIncludePath, stackProjectRoot)

-- | 'ReaderT' + 'IO' monad described here:
-- https://www.fpcomplete.com/blog/2017/07/the-rio-monad
newtype RIO env a = RIO
    { runRIO :: ReaderT env IO a
    } deriving ( Functor, Applicative, Monad, MonadIO
               , MonadReader env, MonadThrow, MonadCatch, MonadMask)

-- | Environment for @importify cache@ command.
data CacheEnvironment = CacheEnvironment
    { -- | Path to local @.importify@ folder
      _pathToImportify :: !(Path Abs Dir)

      -- | Path to GHC .h files
    , _ghcIncludeDir   :: !(Maybe (Path Abs Dir))

      -- | 'True' if unpacked sources should be stored locally as well
    , _saveSources     :: !Bool
    }

makeLensesWith fieldsVerboseLensRules ''CacheEnvironment

type HasPathToImportify env = HasPolyPathToImportify env (Path Abs Dir)
type HasGhcIncludeDir   env = HasPolyGhcIncludeDir   env (Maybe (Path Abs Dir))
type HasSaveSources     env = HasPolySaveSources     env Bool

-- | Getter of @~\/path\/to\/project\/.importify\/symbols@ folder.
pathToSymbols :: HasPathToImportify env => SimpleGetter env (Path Abs Dir)
pathToSymbols = pathToImportify.to (</> symbolsPath)

-- | Run @importify cache@ command. This function takes current
-- project directory and searches for ghc include path.
runCache :: Bool -> RIO CacheEnvironment () -> IO ()
runCache _saveSources cacheAction = do
    projectRoot <- runMaybeT stackProjectRoot
    case projectRoot of
        Nothing              -> return ()  -- error is reported inside 'stackProjectRoot'
        Just projectRootPath -> doInsideDir projectRootPath cacheRunner
  where
    cacheRunner :: IO ()
    cacheRunner = do
        projectPath         <- getCurrentDir
        let _pathToImportify = projectPath </> importifyPath
        _ghcIncludeDir      <- runMaybeT ghcIncludePath
        usingReaderT CacheEnvironment{..} $ runRIO cacheAction
