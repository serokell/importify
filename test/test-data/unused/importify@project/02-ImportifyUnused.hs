module ImportifyUnused where

import           Importify.Cabal (packageDependencies)

main :: IO ()
main = pure ()
