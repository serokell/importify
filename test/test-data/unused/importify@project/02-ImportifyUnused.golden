module ImportifyUnused where


main :: IO ()
main = pure ()
